### PR TITLE
[PyTorch] Store Tensor explicitly in IValue

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50083 [PyTorch] List::operator[] can return const ref for Tensor & string
* #48868 [PyTorch] IValue::toTensor can now return const Tensor&
* **#48824 [PyTorch] Store Tensor explicitly in IValue**
* #49718 [PyTorch] Additional IValue tests

Enables following diff, which will make toTensor() return
`const Tensor&` and allow callers to avoid refcounting overhead.

Differential Revision: [D25324617](https://our.internmc.facebook.com/intern/diff/D25324617/)

### CODE REVIEW

**Review by smessmer — APPROVED** [review]

File: aten/src/ATen/core/ivalue.h [hunk 1]
```diff
@@ -826,13 +871,23 @@ struct CAFFE2_API IValue final {
...
     bool as_bool;
     c10::intrusive_ptr_target* as_intrusive_ptr;
+    at::Tensor as_tensor;
     struct {
       DeviceType type;
       DeviceIndex index;
     } as_device;
+
+    Payload() : as_int(0) {}
+    ~Payload() {}
```
[the remark this answers is not in the document]
Comment (swolchok) [reply aten/src/ATen/core/ivalue.h:881]: Unions with non-POD types in them are a pain. The destructor cannot be defined by default -- do you run ~Tensor() or not? So, we have to define it to do nothing.

File: aten/src/ATen/core/ivalue.h [hunk 2]
```diff
@@ -826,13 +871,23 @@ struct CAFFE2_API IValue final {
...
   };
 
-  IValue(Payload p, Tag t, bool i) : payload(p), tag(t), is_intrusive_ptr(i) {}
+  IValue(const Payload& p, Tag t, bool i) : tag(t), is_intrusive_ptr(i) {
```
[the remark this answers is not in the document]
Comment (swolchok) [reply aten/src/ATen/core/ivalue.h:884]: > Payload has trivial copy/move constructors

Not with `Tensor` in it -- do you run the Tensor copy/move constructors or not? It's not copyable.

File: aten/src/ATen/core/ivalue.h [hunk 3]
```diff
@@ -163,23 +163,61 @@ struct CAFFE2_API IValue final {
...
-    IValue(std::move(rhs)).swap(*this); // this also sets rhs to None
+
+  // Always-inline for performance -- this gets called frequently
+  // inside the core of the static runtime.
+  C10_ALWAYS_INLINE IValue& operator=(IValue&& rhs) & noexcept {
+    if (&rhs == this) {
+      return *this;
+    }
+
+    // Tear down our state.
```
[the remark this answers is not in the document]
Comment (swolchok) [reply aten/src/ATen/core/ivalue.h:193]: Sure, but that's more work for the inliner to get right and these code paths are critical. I can try it.

File: aten/src/ATen/core/ivalue.h [hunk 4]
```diff
@@ -288,7 +326,19 @@ struct CAFFE2_API IValue final {
 
   /// @private [doxygen private]
   void swap(IValue& rhs) noexcept {
-    std::swap(payload, rhs.payload);
+    if (isTensor() && rhs.isTensor()) {
+      std::swap(payload.as_tensor, rhs.payload.as_tensor);
+    } else if (isTensor()) {
+      at::Tensor t = std::move(payload.as_tensor);
```
[the remark this answers is not in the document]
Comment (swolchok) [reply aten/src/ATen/core/ivalue.h:332]: IIUC, you have to construct it in rhs.payload using placement new or it's UB. Skiping the destructor call is legit, and I'll probably try that.

**Comment by swolchok** [comment]
@smessmer could you take another look? I've had to change the approach to improve performance.

**Pull request closed without merging**